File: AED II/algoritmos/mergeSort.js
const Merge = (A, p, q, r) => {
  n1 = q - p + 1; // Numero de elementos do primeiro subVetor
  n2 = r - q; // Numero de elementos do segundo subVetor
  let L = [] // Inicializando os vetores L e R
  let R = []
  for (let i = 0; i < n1; i++) {
    L[i] = A[p + i] // Preenchendo o vetor L com os elementos do primeiro subVetor
  }
  for (let j = 0; j < n2; j++) {
    R[j] = A[q + j + 1] // Preenchendo o vetor R com os elementos do segundo subVetor
  }

  L[n1] = Infinity // Adicionando um valor infinito no fim de L e R
  R[n2] = Infinity
  
  let i = 0
  let j = 0
  for (let k = p; k <= r; k++) {
    if (L[i] <= R[j]) {
      A[k] = L[i]
      i++
    } else {
      A[k] = R[j]
      j++
    }
  }
}


const MergeSort = (A, p, r) => {
  if (p < r) {
    let q = Math.floor((p + r) / 2)
    MergeSort(A, p, q)
    MergeSort(A, q + 1, r)
    Merge(A, p, q, r)
  }
  return A
}


const A = [67,3423,8,987,6543,23254,5568,87675654,332,2435,6789]

console.log(MergeSort(A, 0, A.length - 1))

File: AED II/algoritmos/sorting/merge_sort.py
def merge(A, p, q, r):
    n1 = q - p + 1  # Número de elementos do primeiro sub-vetor
    n2 = r - q  # Número de elementos do segundo sub-vetor
    
    L = []  # Inicializando os vetores L e R
    R = []
    
    for i in range(n1):
        L.append(A[p + i])  # Preenchendo o vetor L com os elementos do primeiro sub-vetor
    
    for j in range(n2):
        R.append(A[q + j + 1])  # Preenchendo o vetor R com os elementos do segundo sub-vetor
    
    L.append(float('inf'))  # Adicionando um valor infinito no fim de L e R
    R.append(float('inf'))
    
    i = 0
    j = 0
    
    for k in range(p, r + 1):
        if L[i] <= R[j]:
            A[k] = L[i]
            i += 1
        else:
            A[k] = R[j]
            j += 1


def merge_sort(A, p, r):
    if p < r:
        q = (p + r) // 2
        merge_sort(A, p, q)
        merge_sort(A, q + 1, r)
        merge(A, p, q, r)
    return A


A = [67, 3423, 8, 987, 6543, 23254, 5568, 87675654, 332, 2435, 6789]

print(merge_sort(A, 0, len(A) - 1))
